reactioncommerce/reaction
# (feat) Allow plugins to override Order Id generation

Resolves #5052 
Impact: **major**  
Type: **feature**

## Issue
Currently we use the Random id to generate Order Ids. A lot of systems have requirements for IDs to be of a different type or in a specific order (e.g. consecutive integers). This allows a plugin to override this by providing a `createOrderReferenceId` and a `createCartReferenceId` function. The default method of using `Random.id()` remains the same.

It also adds a `referenceId` to the cart

## Solution
Allow for a `createOrderId` function to be provided by a plugin and use that instead of Random.id

## Breaking changes
None

## Testing
1. Create an order and verify it has an order id
1. Create a test plugin in `imports/plugins/custom/fake` with just this file


register.js
```
import Reaction from "/imports/plugins/core/core/server/Reaction";
import _ from "lodash";

function createNumericId() {
  console.log("createdNumericId");
  return _.random(0, 1000).toString();
}

Reaction.registerPackage({
  label: "Fake",
  name: "reaction-fake",
  icon: "fa fa-fake",
  autoEnable: true,
  functionsByType: {
    createCartReferenceId: [createNumericId]
  }
});
```

3. Restart Reaction
4. Create an order
5. Verify that the order Id is a number between 0 and 1000.
